### DISCUSSION TITLE
Created documentation of debugging MeshCentral, kindly asking for help to understand win-terminal.js code

### DISCUSSION THREAD

**Discussion by MartinMa** [Ylianst/MeshCentral#4898] [Q&A]
# A – Goal

1. We (@schories and I) want to fix the issue: https://github.com/Ylianst/MeshCentral/issues/4327
2. We want to document - initially within this discussion - all debug information in one place, so that you or search engines can easily point other users/developers to it.

# B – How we debug

## a) According to the information on the internet (documentation, former issues, your comments on reddit, etc. ) we are using the following approach to debug:

1. On the Agent PC:
    Place - original or modified - `win-terminal.js` into the MeshAgent directory on Windows 10 LTSB 1607.

2. On the Agent PC:
    Run the agent with `MeshAgent.exe run` within an elevated/local-administrator CMD box.

3. On the Management PC:
    Opening MeshCentral in the Webbrowser with `&trace=1` added to the URL.

4. On the Management PC:
    In the (MeshCentral) Agent View switch to Console tab.
    Execute `setdebug 2` to enable log output within the Console window.
5. On the Management PC:
    Invoke the terminal from the Terminal tab.

## b) We are impacted by a few issues with the above debugging procedure:

When using `console.log` in the `win-terminal.js` 

- we have experienced missing log messages in the Console tab. 
- It is unclear to us, whether the messages are not created or send by the agent or are lost within the MeshCentral server.

Moreover we tried `duktape-debugger.js`: 

- However, neither in current versions of Edge, Chrome or Firefox breakpoints did work although shown/set correctly. We couldn't step through the actual code. 
- Clearing breakpoints would lead to those breakpoints visually reappearing when reloaded. 
- As a result this approach is unusable, at least on our systems.

# C – Our basic understanding how the remote terminal is implemented:

## On the MeshCentral side 

- [ANSI escape code](https://en.wikipedia.org/wiki/ANSI_escape_code) messages are received from the agent (as we use MeshCentral server-based redirecion; no direct network connection between Agent PC and Management PC)
- and passed to [xterm.js](http://xtermjs.org/) to render the terminal view

## On the Agent side

- `GetConsoleWindow()` is called to retrieve the window handle of the console associated with the calling process.
- Optional: If the former call fails a call to `AllocConsole()` ensures, that a new console is allocated and associated to the calling process.
  A process can have at most one console associated to it.
- `GetStdHandle()` is called twice to receive the standard input and output handles of the console.
- `SetConsoleWindowInfo()` is called to determine the size of the console window to a hardcoded size of 80 columns and 25 rows.

- `SetConsoleScreenBufferSize` is called to determine the screen buffer size of the console and make it the same size as the console window (80 columns and 25 rows).

- A "clear screen" is perfomed involving various Windows API calls.

- `SetWinEventHook()` is called to start observing events from `EVENT_CONSOLE_CARET` to `EVENT_CONSOLE_END_APPLICATION` for all processes and threads (because the parameters for `idProcess` and `idThread` are both set to 0) along with the flags `WINEVENT_OUTOFCONTEXT | WINEVENT_SKIPOWNPROCESS`. The hook function is called asynchronous across process boundaries. We will receive the messages in sequential order, but not at the very moment they appear. More on that later.

- After that a simple message loop is rolled out using `GetMessage()`, `TranslateMessage()` and `DispatchMessage()`. On `WM_QUIT` the hook is removed by calling `UnhookWinEvent()` and the process is terminated by calling `TerminateProcess()`.

- After setting up the message loop `CreateProcess()` is called to create a new process with the flag `CREATE_NEW_PROCESS_GROUP`, which means it is inheriting its parent's (calling process) console.

- The hook function, which was attached earlier, basically handles the seven events `EVENT_CONSOLE_CARET`, `EVENT_CONSOLE_UPDATE_REGION`, `EVENT_CONSOLE_UPDATE_SIMPLE`, `EVENT_CONSOLE_UPDATE_SCROLL`, `EVENT_CONSOLE_LAYOUT`, `EVENT_CONSOLE_START_APPLICATION` and `EVENT_CONSOLE_END_APPLICATION`.

# D – Our 3 questions

1. Debug: We kindly ask if there are better/other ways to debug running module code within duktape?

2. Code: While handling `EVENT_CONSOLE_UPDATE_REGION` *and* the last line (25) of the console screen buffer is reached, the actual issue occurs:

   - Although the buffer contents in the actual CMD window of the Agent is correctly shifted up by one row - all following region update events are causing the overwriting of the same last line in xterm.js counterpart on the MeshCentral website. 
     The xterm.js "shift up information" is simply never detected/sent.
   - Additional observations: 
     - As console window size and screen buffer size are set identically, the event `EVENT_CONSOLE_UPDATE_SCROLL` is never invoked.
     - Increasing the screen buffer to e.g. 100 rows causes the problem to occur after 100 lines, instead of 25. 

   Please help us understand at which point the shifting is supposed to occur? How can we detect/debug the shifting moment?

3. Code: Kindly give us some hint, what the benefit of / reason for using `peek` is in the following context: 

- https://github.com/Ylianst/MeshCentral/blob/bd57601cd8a0f6c069f9d5e53ca50cbb1f1071ce/agents/modules_meshcore/win-terminal.js#L242-L254

- https://github.com/Ylianst/MeshCentral/blob/bd57601cd8a0f6c069f9d5e53ca50cbb1f1071ce/agents/modules_meshcore/win-terminal.js#L613-L618

**Comment by silversword411** [comment]
Anything we can do here to enhance debugging/understanding would be good

https://ylianst.github.io/MeshCentral/meshcentral/debugging/
